port/cpl_vsil_s3.cpp

```diff
@@ -2034,45 +2027,88 @@ int VSIS3WriteHandle::Close()
     return nRet;
 }
 
+/************************************************************************/
+/*                          CreateWriteHandle()                         */
+/************************************************************************/
+
+VSIVirtualHandleUniquePtr
+VSIS3FSHandler::CreateWriteHandle(const char *pszFilename,
+                                  CSLConstList papszOptions)
+{
+    auto poHandleHelper =
+        CreateHandleHelper(pszFilename + GetFSPrefix().size(), false);
+    if (poHandleHelper == nullptr)
+        return nullptr;
+    auto poHandle = cpl::make_unique<VSIS3WriteHandle>(
+        this, pszFilename, poHandleHelper, false, papszOptions);
+    if (!poHandle->IsOK())
+    {
+        return nullptr;
+    }
+    return VSIVirtualHandleUniquePtr(poHandle.release());
+}
+
 /************************************************************************/
 /*                                Open()                                */
 /************************************************************************/
 
-VSIVirtualHandle *VSIS3FSHandler::Open(const char *pszFilename,
-                                       const char *pszAccess, bool bSetError,
-                                       CSLConstList papszOptions)
+VSIVirtualHandle *VSICurlFilesystemHandlerBaseWritable::Open(
+    const char *pszFilename, const char *pszAccess, bool bSetError,
+    CSLConstList papszOptions)
 {
     if (!STARTS_WITH_CI(pszFilename, GetFSPrefix()))
         return nullptr;
 
-    if (strchr(pszAccess, 'w') != nullptr || strchr(pszAccess, 'a') != nullptr)
+    if (strchr(pszAccess, '+'))
     {
-        if (strchr(pszAccess, '+') != nullptr &&
-            !SupportsRandomWrite(pszFilename, true))
+        if (!SupportsRandomWrite(pszFilename, true))
         {
             CPLError(CE_Failure, CPLE_AppDefined,
-                     "w+ not supported for /vsis3, unless "
-                     "CPL_VSIL_USE_TEMP_FILE_FOR_RANDOM_WRITE is set to YES");
+                     "%s not supported for %s, unless "
+                     "CPL_VSIL_USE_TEMP_FILE_FOR_RANDOM_WRITE is set to YES",
+                     pszAccess, GetFSPrefix().c_str());
             errno = EACCES;
             return nullptr;
         }
 
-        VSIS3HandleHelper *poS3HandleHelper = VSIS3HandleHelper::BuildFromURI(
-            pszFilename + GetFSPrefix().size(), GetFSPrefix().c_str(), false);
-        if (poS3HandleHelper == nullptr)
-            return nullptr;
-        VSIS3WriteHandle *poHandle = new VSIS3WriteHandle(
-            this, pszFilename, poS3HandleHelper, false, papszOptions);
-        if (!poHandle->IsOK())
+        const std::string osTmpFilename(CPLGenerateTempFilename(nullptr));
```

Comment:
Could a user set `CPL_TMPDIR=/vsimem/` to write the tempfile to memory with a goal of lower IO latency assuming there is enough free RAM available?

https://github.com/OSGeo/gdal/blob/944a839bc0b8847585e017792985e04aced58e6e/port/cpl_path.cpp#L1115

Reply:
> Could a user set `CPL_TMPDIR=/vsimem/` to write the tempfile to memory with a goal of lower IO latency assuming there is enough free RAM available?

yes, should work